### ISSUE TITLE
vertical motion (j or k commands) don't seem to act correctly with vimtex?  This may be recent...

### ISSUE THREAD

**Issue by tunaboy** [lervag/vimtex#1158]
When I load vimtex, with a stripped down `.vimrc` and `.vim` directory, I continue to have an issue where the vertical motion "j" or "k" commands act anomolously. With the cursor in column 5 (say) if I then hit `j` multiple times to go down, if I pass through an empty line, the cursor stays in the first column rather than jumping back to column 5 after I'm back on a line with at least 5 columns.

It's seems like a small thing, but it's becoming an issue.

My [stripped down `.vimrc`](https://github.com/lervag/vimtex/files/2173588/default.vimrc.gz):

```vim
execute pathogen#infect()
syntax on
filetype plugin indent on`
```

My `.vim` directory (also stripped down now) has simply the bundle directory (where I've cloned the vimtex repository) and the autoload directory (where pathogen.vim is the only file).

A [mwe.tex.gz](https://github.com/lervag/vimtex/files/2173552/mwe.tex.gz) file:

```tex
\documentclass{minimal}
\begin{document}

Hello World!

\end{document}
```

The issue is simple.  Open the mwe, and move the cursor to the 5th column on the first line (the character u).  now hit j 3 times so that you are on the line that is "Hello World!".  At this point, the cursor should be on the first "o" of "Hello World!", and instead it is on the first column (with H).  

Simply rename the file to mwe.txt, and redo the exercise, and the cursor will end up in exactly the correct spot (the first "o" of "Hello World!").

I'm running vim 7.4.1009  (on Centos 7).  I believe I see the same behaviour on Suse machines, which are probably running a different vim version (but likely close).  

That the behaviour is corrected by simply changing the filename to txt (so that vimtex is not involved) is pretty indicative that it is vimtex that is causing this problem. 

I have enjoyed this software, and I have been getting away from latex-suite, but this is a problem for me.  I'm hoping this is a quick fix, or I need to upgrade my vim, or I'm doing something wrong I am not seeing.  I also suspect this might be a recent thing, as I would have noticed earlier, but I can't be sure.

Thanks much,
Mike

**Comment by lervag** [comment]
Hi Mike. Could you please reproduce this with a full minimal `.vimrc` file, e.g.

```vim
set nocompatible
let &rtp = '~/.vim/bundle/vimtex,' . &rtp
let &rtp .= ',~/.vim/bundle/vimtex/after'
filetype plugin indent on
syntax enable
```

I.e., no need for `pathogen`; the vimrc file is barely longer. Just make sure that the paths are correct.

Pro tip: You do not need to strip down your personal vim configuration for this. It is enough to save the minimal vimrc file to `minivimrc`, then do: `vim -u minivimrc mwe.tex`.

For your info: I can not reproduce the problem with Vim or neovim on my
machine. You say that you do not have the problem when you change the filename.
But now instead, can you check if the problem persists if you remove one of the
two `&rtp` lines above, e.g. with the following minimal vimrc file:

```vim
set nocompatible
" Either
"let &rtp = '~/.vim/bundle/vimtex,' . &rtp
let &rtp .= ',~/.vim/bundle/vimtex/after'

" Or
let &rtp = '~/.vim/bundle/vimtex,' . &rtp
"let &rtp .= ',~/.vim/bundle/vimtex/after'

" Or
"let &rtp = '~/.vim/bundle/vimtex,' . &rtp
"let &rtp .= ',~/.vim/bundle/vimtex/after'

filetype plugin indent on
syntax enable
```

Please ensure that you are on the latest version of vimtex as well. I assume
you are based on your initial info, but still.

**Comment by tunaboy** [comment]
OK, this is instructive.
I used your minimal .vimrc, and in the manner you suggested.  I'm now on my work machine, and running vim version 7.4.160 (which I suspect is even older).  This is a CENTOS 7 machine also, and I can try the SuSe machine, on which I think I have the stock vim and also one I compiled a while back which may be more recent than these stock versions.

So, with the minimal .vimrc, and a .vim that is only the bundle/vimtex directory (whcih I dropped into and did a git update, which did update some files on my work machine) I tried out the mwe, and it continued to _not_ work correctly.  
 I then tried the top version, and it works correctly - (which would imply the bad magic is not occuring in the "after" coding)
So 1st alternative works:
 ``set nocompatible
"let &rtp = '~/.vim/bundle/vimtex,' . &rtp
let &rtp .= ',~/.vim/bundle/vimtex/after'
filetype plugin indent on
syntax enable
-------------------------------------------------
The next .vimrc, where the "after" is commented out, but the main one is not - problem came back.  

set nocompatible
let &rtp = '~/.vim/bundle/vimtex,' . &rtp
"let &rtp .= ',~/.vim/bundle/vimtex/after'
filetype plugin indent on
syntax enable
-----------------------------------------------------------------------------
And the final one, where both are commented out works correctly also

`set nocompatible
"let &rtp = '~/.vim/bundle/vimtex,' . &rtp
"let &rtp .= ',~/.vim/bundle/vimtex/after'
filetype plugin indent on
syntax enable
 `
---------------------------------------------------------------------------------------------
OK, now it got interesting.  I got to the SuSe machine, where I have my own compiled version of vim, and there is no problem.  I did the same exercise, with your stripped down .vimrc, and a .vim that was only the bundle directory with vimtex, and it works perfectly.  No problems.  This is vim 8.0.799, and I think I have found my issue.

I'll compile my own vim on my work machine, and see if that fixes it, but it certainly seems to be in the vim and not in the vimtex.  

Thanks much for your help in this.  I'll get back to you to let you know if that fixes it.  I use tex to do plotting, (with pgfplots) and this becomes a real productivity issue.  This will fix it, and I'll also have a more up to date vim (hopefully)...

Thanks much,
Mike

Mike

**Comment by tunaboy** [comment]
Sorry for the formatting, not sure what I did, but wanted to give you the info sooner rather than later.  
Will be back to say whether the "custom vim" fixes the issue.

Thanks 
Mike

**Comment by tunaboy** [comment]
And, now having compiled a new vim (version 8.1), the problem is fixed.  

Not sure why the 7.4 versions would have problems, but that does seem to be the issue.

Thanks much, I now have things working nicely again.  I would not have suspected that the standard vim on Centos would be an issue.

**Comment by lervag** [comment]
So, if I understand correctly, you say that you get the problem when you use Vim 7.4.160 with the following vimrc file:

```vim
set nocompatible
let &rtp = '~/.vim/bundle/vimtex,' . &rtp
"let &rtp .= ',~/.vim/bundle/vimtex/after'
filetype plugin indent on
syntax enable
```

Also, that the problem is gone with a newer Vim version. I have a suspicion that it is due to the highlighting of matching parantheses (and similar). Could you try to add the option `let g:vimtex_matchparen_enabled = 0` to your vimrc and see if it helps for the old Vim version?

**Comment by tunaboy** [comment]
First, thanks much for cleaning up my post so it's readable.  I could not find the character/code to delete that was goofing up the post.  It looks great now.

So, on the second point, I just added

let g:vimtex_matchparen_enabled = 0

to my full blown .vimrc and .vim on the 7.4.1099 version, and it entirely fixes the problem.  

Just for completeness, I swapped out my .vim for one that only has bundle, and first used the minimal .vimrc you suggested at first (w/o the final suggestion), and reproduced the behavior I got at work with the 7.4 version there.  I then added the `let g:vimtex_matchparen_enabled=0` to that minimal .vimrc, and again, it fixes the problem.  

So, yes, that does fix the problem with the older vim version.

**Comment by tunaboy** [comment]
And the same applies at work, with the let g:vimtex_matchparen_enabled = 0 fixing the issue using the old version of vim (7.4.160)

**Comment by lervag** [comment]
> First, thanks much for cleaning up my post so it's readable. I could not find the character/code to delete that was goofing up the post. It looks great now.

No problem. If you want, you can read about formatting of posts [here](https://guides.github.com/features/mastering-markdown/).

> And the same applies at work, with the `let g:vimtex_matchparen_enabled = 0` fixing the issue using the old version of vim (7.4.160).

Thanks, that confirms my suspicion. I'll look into it to see if I can fix the problem for old Vim's, but in the meantime, you could leave the highlighting of by default with the setting. I find the highlighting is quite useful, but not really that important.

**Comment by lervag** [comment]
Ok, so, the problem is that the code that does the highlighting needs to move the cursor. This relies on a feature that was provided in Vim 7.4.313, i.e. `getcurpos()`, which includes information about which vertical column we want in case the number of columns is smaller than the "previous" line. To be brief, it seems to be impossible to make this work properly for older Vim.

A workaround is to add something like this in your vimrc:

```vim
if !exists('*getcurpos')
  let g:vimtex_matchparen_enabled = 0
endif
```

This will disable the matchparen feature if the required feature is not available.

However, it seems there might be some more changes added between patch 313 and sometime after patch 1099, since things do not work even when patch 313 is included. Perhaps the `cursor()` function was not updated. You could also use something like this:

```vim
if v:version < 800
  let g:vimtex_matchparen_enabled = 0
endif
```

**Comment by tunaboy** [comment]
Just using the newer vim works for me.  If I am forced to, I'll just use the 
   let g:vimtex_matchparen_enabled = 0
if I'm caught w/o a newer vim available.  

I did add 
 ` if v:version < 800
  let g:vimtex_matchparen_enabled = 0
endif`
to the end of my .vimrc, and it does fix the  problem with the older version of gvim/vim, but now that I've got 8.1, I don't think I'll go back.  

Thank you very much!  

Mike
